cloudsmith/resource_repository_geo_ip_rules.go

```diff
@@ -0,0 +1,184 @@
+const CidrDeny string = "cidr_deny"
+const CountryCodeAllow string = "country_code_allow"
+const CountryCodeDeny string = "country_code_deny"
+
+func resourceRepositoryGeoIpRulesCreate(d *schema.ResourceData, m interface{}) error {
+	pc := m.(*providerConfig)
+
+	namespace := requiredString(d, Namespace)
+	repository := requiredString(d, Repository)
+
+	// Ensure that Geo/IP rules are enabled for the Repository
+	req := pc.APIClient.ReposApi.ReposGeoipEnable(pc.Auth, namespace, repository)
+	_, err := pc.APIClient.ReposApi.ReposGeoipEnableExecute(req)
+	if err != nil {
+		return err
+	}
+
+	// The actual "create" is just the same as "update" for this resource.
+	return resourceRepositoryGeoIpRulesUpdate(d, m)
+}
+
+func resourceRepositoryGeoIpRulesRead(d *schema.ResourceData, m interface{}) error {
+	pc := m.(*providerConfig)
+
+	namespace := requiredString(d, Namespace)
+	repository := requiredString(d, Repository)
+
+	req := pc.APIClient.ReposApi.ReposGeoipRead(pc.Auth, namespace, repository)
+
+	geoIpRules, resp, err := pc.APIClient.ReposApi.ReposGeoipReadExecute(req)
+	if err != nil {
+		if is404(resp) {
+			d.SetId("")
+			return nil
+		}
+
+		return err
+	}
+
+	cidr := geoIpRules.GetCidr()
+	countryCode := geoIpRules.GetCountryCode()
+
+	_ = d.Set(CidrAllow, flattenStrings(cidr.GetAllow()))
+	_ = d.Set(CidrDeny, flattenStrings(cidr.GetDeny()))
+	_ = d.Set(CountryCodeAllow, flattenStrings(countryCode.GetAllow()))
+	_ = d.Set(CountryCodeDeny, flattenStrings(countryCode.GetDeny()))
+
+	// namespace and repository are not returned from the read
+	// endpoint, so we can use the values stored in resource state. We rely on
+	// ForceNew to ensure if either changes a new resource is created.
+	_ = d.Set(Namespace, namespace)
+	_ = d.Set(Repository, repository)
+
+	return nil
+}
+
+func resourceRepositoryGeoIpRulesUpdate(d *schema.ResourceData, m interface{}) error {
+	pc := m.(*providerConfig)
+
+	namespace := requiredString(d, Namespace)
+	repository := requiredString(d, Repository)
+
+	req := pc.APIClient.ReposApi.ReposGeoipUpdate(pc.Auth, namespace, repository)
+	req = req.Data(cloudsmith.ReposGeoipRead200Response{
+		CountryCode: &cloudsmith.ReposGeoipRead200ResponseCountryCode{
+			Allow: expandStrings(d, CountryCodeAllow),
+			Deny:  expandStrings(d, CountryCodeDeny),
+		},
+		Cidr: &cloudsmith.ReposGeoipRead200ResponseCountryCode{
+			Allow: expandStrings(d, CidrAllow),
+			Deny:  expandStrings(d, CidrDeny),
+		},
+	})
+
+	_, err := pc.APIClient.ReposApi.ReposGeoipUpdateExecute(req)
+	if err != nil {
+		return err
+	}
+
+	d.SetId(fmt.Sprintf("%s_%s_geo_ip_rules", namespace, repository))
+
+	return resourceRepositoryGeoIpRulesRead(d, m)
+}
+
+func resourceRepositoryGeoIpRulesDelete(d *schema.ResourceData, m interface{}) error {
+	pc := m.(*providerConfig)
+
+	namespace := requiredString(d, "namespace")
+	repository := requiredString(d, "repository")
+
+	// There isn't a DELETE endpoint, so just update the rules to be empty.
```

Comment:
do we need to disable the geoip enforcement in here too?

Reply:
I don't mind doing that, I just didn't think it would be necessary given that it makes no material difference if you don't have any rules in place. Your call!